src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,326 @@
+            // Now let's parse the file name.
+            string updateFileName = Path.GetFileName(updateFilePath);
+            int dateStartIndex = updateFileName.LastIndexOf('_') + 1;
+
+            bool success = DateTime.TryParse(
+                updateFileName.AsSpan().Slice(dateStartIndex),
+                CultureInfo.InvariantCulture,
+                DateTimeStyles.AssumeLocal,
+                out lastUpdateDate);
+
+            if (success)
+            {
+                int versionStartIndex = updateFileName.IndexOf('_') + 2;
+                int versionLength = dateStartIndex - versionStartIndex - 1;
+                string versionString = updateFileName.Substring(versionStartIndex, versionLength);
+                success = SemanticVersion.TryParse(versionString, out lastUpdateVersion);
+            }
+
+            if (!success)
+            {
+                updateFilePath = null;
+                lastUpdateVersion = null;
+                lastUpdateDate = default;
+            }
+
+            return success;
+        }
+
+        private static async Task<Release> QueryNewReleaseAsync(SemanticVersion baselineVersion)
+        {
+            bool notPreRelease = string.IsNullOrEmpty(PSVersionInfo.PSCurrentVersion.PreReleaseLabel);
+            string queryUri = notPreRelease ? LatestReleaseUri : Last3ReleasesUri;
+
+            using (HttpClient client = new HttpClient())
+            {
+                string userAgent = string.Format(CultureInfo.InvariantCulture, "PowerShell {0}", PSVersionInfo.GitCommitId);
+                client.DefaultRequestHeaders.Add("User-Agent", userAgent);
+                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+                var response = await client.GetAsync(queryUri);
+                response.EnsureSuccessStatusCode();
+
+                using (var stream = await response.Content.ReadAsStreamAsync())
+                using (var reader = new StreamReader(stream))
+                using (JsonReader jsonReader = new JsonTextReader(reader))
+                {
+                    Release releaseToReturn = null;
+                    var settings = new JsonSerializerSettings() { DateParseHandling = DateParseHandling.None };
+                    var serializer = JsonSerializer.Create(settings);
+
+                    if (notPreRelease)
+                    {
+                        var release = serializer.Deserialize<JObject>(jsonReader);
+                        var tagName = release["tag_name"].ToString();
+                        var version = SemanticVersion.Parse(tagName.Substring(1));
+
+                        if (version > baselineVersion)
+                        {
+                            var publishAt = release["published_at"].ToString();
+                            releaseToReturn = new Release(publishAt, tagName);
+                        }
+                    }
+                    else
```

Comment:
Could avoid the `else` here by returning early -- prevents me needing to read to the bottom to see what the shared state is at the end